# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent


### 1. iNaturalist statuses

In [133]:
import pandas as pd
import sys
import os
projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf

sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
# taxastatus.head(3)
taxastatus

id taxon_id  user_id place_id source_id  \
0     166449    38493  1138587     7830             
1     234788   918383   702203     9994             
2     234789   918383   702203     7308             
3     166416  1033183  3669610     6825             
4     180721  1247288   222137     6825             
...      ...      ...      ...      ...       ...   
3448  165697  1182117  3669610    73684             
3449  130966   508985                               
3450  161226    50744   516268                      
3451  162783   924263   764897                      
3452  152695   779693   708886              16653   

                                              authority  \
0                    Flora and Fauna Guarantee Act 1988   
1                             Atlas of Living Australia   
2                             Atlas of Living Australia   
3                  NSW Office of Environment & Heritage   
4           NSW Threatened Species Scientific Committee   
...                                                 ...   
3448                              Australian Government   
3449                                                      
3450          DAWE Species Profile and Threats Database   
3451  Department of Biodiversity, Conservation and A...   
3452                                           QLD DEHP   

                     status  \
0                        CR   
1                       NT    
2                        LC   
3                        EN   
4                        vu   
...                     ...   
3448                     CR   
3449  critically endangered   
3450                     CR   
3451                     EN   
3452             endangered   

                                                    url description  \
0                                                                     
1     https://bie.ala.org.au/species/https://id.biod...               
2     https://bie.ala.org.au/species/https://id.biod...               
3     https://www.environment.nsw.gov.au/threateneds...               
4     https://www.environment.nsw.gov.au/topics/anim...               
...                                                 ...         ...   
3448  http://www.environment.gov.au/biodiversity/thr...               
3449  http://environment.gov.au/cgi-bin/sprat/public...               
3450  https://www.environment.gov.au/cgi-bin/sprat/p...               
3451  https://www.dpaw.wa.gov.au/images/documents/pl...               
3452  https://data.qld.gov.au/dataset/conservation-s...               

     geoprivacy  ...          genus specificEpithet infraspecificEpithet  \
0      obscured  ...      Eulamprus      kosciuskoi                        
1                ...   Chiloschista     phyllorhiza                        
2                ...   Chiloschista     phyllorhiza                        
3      obscured  ...       Eidothea      hardeniana                        
4      obscured  ...     Pomaderris         bodalla                        
...         ...  ...            ...             ...                  ...   
3448             ...    Achyranthes    margaretarum                        
3449   obscured  ...  Lichenostomus        melanops             cassidix   
3450   obscured  ...                                                       
3451   obscured  ...         Reedia       spathacea                        
3452   obscured  ...   Selenocosmia       crassipes                        

                  modified                   scientificName   taxonRank  \
0     2021-03-01T10:35:01Z             Eulamprus kosciuskoi     species   
1     2022-01-08T03:30:36Z         Chiloschista phyllorhiza     species   
2     2022-01-08T03:30:36Z         Chiloschista phyllorhiza     species   
3     2021-02-22T07:21:17Z              Eidothea hardeniana     species   
4     2021-08-27T06:18:35Z               Pomaderris bodalla     species   
...                    ...                              ...         ...   


In [134]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("NSW|New South Wales", ".nsw.")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id  user_id place_id source_id  \
3       166416  1033183  3669610     6825             
2656    165059  1054498    58320     6825             
477     264941  1061113  3669610     6825             
371     180988  1070573  3669610     6825             
2514    160581  1076814   990532     6825             
...        ...      ...      ...      ...       ...   
1111    152247   906670   708886     6825     16650   
2637    164342   913024    58320     6825             
7       167825   953250   702203     6825             
836     155144   966856   708886     6825     16650   
2651    164615   973457   222137     6825             

                                              authority      status  \
3                  NSW Office of Environment & Heritage          EN   
2656               NSW Office of Environment & Heritage  Vulnerable   
477   New South Wales Biodiversity Conservation Act ...          EX   
371                NSW Office of Environment & Heritage          EN   
2514               NSW Office of Environment & Heritage  Vulnerable   
...                                                 ...         ...   
1111               NSW Office of Environment & Heritage  Endangered   
2637               NSW Office of Environment & Heritage  Vulnerable   
7     New South Wales Office of Environment and Heri...          VU   
836                NSW Office of Environment & Heritage  endangered   
2651        NSW Threatened Species Scientific Committee  Endangered   

                                                    url       description  \
3     https://www.environment.nsw.gov.au/threateneds...                     
2656  https://www.environment.nsw.gov.au/threatenedS...                     
477   https://bie.ala.org.au/species/https://id.biod...  Presumed Extinct   
371   https://www.environment.nsw.gov.au/threateneds...                     
2514  https://www.environment.nsw.gov.au/threateneds...                     
...                                                 ...               ...   
1111  https://www.environment.nsw.gov.au/resources/a...                     
2637  https://www.environment.nsw.gov.au/threatenedS...                     
7     https://www.environment.nsw.gov.au/threatenedS...        Vulnerable   
836   https://www.environment.nsw.gov.au/resources/a...                     
2651  https://www.environment.nsw.gov.au/topics/anim...                     

     geoprivacy  ...         genus specificEpithet infraspecificEpithet  \
3      obscured  ...      Eidothea      hardeniana                        
2656   obscured  ...  Prostanthera  cryptandroides       cryptandroides   
477        open  ...        Leuzea       australis                        
371              ...        Rotala      tripartita                        
2514   obscured  ...   Pterostylis       nigricans                        
...         ...  ...           ...             ...                  ...   
1111   obscured  ...     Caladenia        concolor                        
2637             ...     Symplocos     baeuerlenii                        
7                ...     Pultenaea          glabra                        
836    obscured  ...        Pteris   platyzomopsis                        
2651   obscured  ...       Calotis       pubescens                        

                  modified                              scientificName  \
3     2021-02-22T07:21:17Z                         Eidothea hardeniana   
2656  2020-12-17T03:53:45Z  Prostanthera cryptandroides cryptandroides   
477   2022-06-11T11:38:00Z                            Leuzea australis   
371   2021-09-21T05:48:55Z                           Rotala tripartita   
2514  2022-07-12T14:19:22Z                       Pterostylis nigricans   
...                    ...                                         ...   
1111  2022-06-10T07:36:31Z                          Caladenia concolor   
2637  2022-03-14T15:27:03Z                       Symplocos baeuerlenii   

### 2. iNaturalist taxonomy

In [135]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists

Sensitive list: `geoprivacy` = `obscured`
overrides


In [136]:
sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr487?max=10000&includeKVP=true")
sensitivelist = lf.kvp_to_columns(sensitivelist)
sensitivelist['name'] = sensitivelist['name'].str.replace('subsp. ', '', regex=False)
sensitivelist = sensitivelist.rename(columns={'T S Profile I D':'bionetTaxonID'})
sensitivelist['bionet_geoprivacy'] = 'obscured'
sensitivelist

id                           name                     commonName  \
0   4040135       Callocephalon fimbriatum             Gang-gang Cockatoo   
0   4040193                  Ninox strenua                   Powerful Owl   
0   4040029               Tyto tenebricosa                      Sooty Owl   
0   4040138                Ninox connivens                    Barking Owl   
0   4040203      Hoplocephalus bungaroides             Broad-headed Snake   
..      ...                            ...                            ...   
0   4040083  Anthosachne kingiana kingiana     Phillip Island Wheat Grass   
0   4040198                Caladenia tensa  Erect Greencomb Spider Orchid   
0   4040162           Caladenia atroclavia    Black-clubbed Spider-orchid   
0   4040056           Prasophyllum sandrae       Majors Creek Leek Orchid   
0   4040139         Backhousia subargentea                 Giant Ironwood   

                          scientificName  \
0               Callocephalon fimbriatum   
0            Ninox (Rhabdoglaux) strenua   
0                       Tyto tenebricosa   
0         Ninox (Hieracoglaux) connivens   
0              Hoplocephalus bungaroides   
..                                   ...   
0   Anthosachne kingiana subsp. kingiana   
0                        Caladenia tensa   
0                   Caladenia atroclavia   
0                   Prasophyllum sandrae   
0                 Backhousia subargentea   

                                                 lsid dataResourceUid  \
0   https://biodiversity.org.au/afd/taxa/6c646af8-...           dr487   
0   https://biodiversity.org.au/afd/taxa/d1c5dee0-...           dr487   
0   https://biodiversity.org.au/afd/taxa/645b287c-...           dr487   
0   https://biodiversity.org.au/afd/taxa/bd332ca4-...           dr487   
0   https://biodiversity.org.au/afd/taxa/f579a6fc-...           dr487   
..                                                ...             ...   
0   https://id.biodiversity.org.au/taxon/apni/5129...           dr487   
0   https://id.biodiversity.org.au/taxon/apni/5139...           dr487   
0   https://id.biodiversity.org.au/taxon/apni/5139...           dr487   
0                                       ALA_DR487_196           dr487   
0    https://id.biodiversity.org.au/node/apni/2888873           dr487   

                                            kvpValues   taxonRank   kingdom  \
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
..                                                ...         ...       ...   
0   [{'key': 'taxonRank', 'value': 'Subspecies'}, ...  Subspecies   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   

       class  ... protectedInNSW sensitivityClass bionetTaxonID  \
0       Aves  ...           True       Category 3       10975.0   
0       Aves  ...           True       Category 3       10562.0   
0       Aves  ...           True       Category 3       10821.0   
0       Aves  ...           True       Category 3       10561.0   
0   Reptilia  ...           True       Category 2       10413.0   
..       ...  ...            ...              ...           ...   
0      Flora  ...          False       Category 3       20148.0   
0      Flora  ...           True       Category 2           NaN   
0      Flora  ...           True       Category 2       20365.0   
0      Flora  ...          False       Category 2       106

In [137]:
def gpcondition(x):
    if x=='Category 1':
        return "private"
    elif x== 'Category 2' or x=='Category 3':
        return "obscured"
    else:
        return 'open'

In [138]:
conservationlist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr650?max=10000&includeKVP=true")
conservationlist = lf.kvp_to_columns(conservationlist)
conservationlist['name'] = conservationlist['name'].str.replace('subsp. ', '', regex=False)
conservationlist = conservationlist.rename(columns={'T S Profile I D':'bionetTaxonID'})
conservationlist['bionet_geoprivacy'] = conservationlist['sensitivityClass'].apply(gpcondition)
conservationlist

id                           name             commonName  \
0   4028582                    Delma impar       Many-lined Delma   
0   4028454       Callocephalon fimbriatum     Gang-gang Cockatoo   
0   4028239            Cacophis harriettae    White-crowned Snake   
0   4028340        Litoria booroolongensis        Booroolong Frog   
0   4028512            Anthochaera phrygia      Regent Honeyeater   
..      ...                            ...                    ...   
0   4028492    Muehlenbeckia sp. Mt Norman      Scrambling Lignum   
0   4028995             Lobelia claviflora                   None   
0   4029167    Fontainea sp. Coffs Harbour                   None   
0   4029112           Melichrus gibberagee  Narrow-leaf Melichrus   
0   4028270  Pittosporum sp. Coffs Harbour                   None   

                      scientificName  \
0                        Delma impar   
0           Callocephalon fimbriatum   
0                Cacophis harriettae   
0            Litoria booroolongensis   
0   Anthochaera (Xanthomyza) phrygia   
..                               ...   
0        Muehlenbeckia sp. Mt Norman   
0                 Lobelia claviflora   
0        Fontainea sp. Coffs Harbour   
0               Melichrus gibberagee   
0      Pittosporum sp. Coffs Harbour   

                                                 lsid dataResourceUid  \
0   https://biodiversity.org.au/afd/taxa/51c700dd-...           dr650   
0   https://biodiversity.org.au/afd/taxa/6c646af8-...           dr650   
0   https://biodiversity.org.au/afd/taxa/ace6534f-...           dr650   
0   https://biodiversity.org.au/afd/taxa/6377e038-...           dr650   
0   https://biodiversity.org.au/afd/taxa/31869a0e-...           dr650   
..                                                ...             ...   
0                                      ALA_DR650_1014           dr650   
0                                      ALA_DR650_1015           dr650   
0                                      ALA_DR650_1016           dr650   
0                                      ALA_DR650_1017           dr650   
0                                      ALA_DR650_1018           dr650   

                                            kvpValues taxonRank   kingdom  \
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
..                                                ...       ...       ...   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
0   [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   

       class  ...           sourceStatus protectedInNSW sensitivityClass  \
0   Reptilia  ...             Vulnerable           True    Not Sensitive   
0       Aves  ...             Vulnerable           True       Category 3   
0   Reptilia  ...             Vulnerable           True    Not Sensitive   
0   Amphibia  ...             Endangered           True    Not Sensitive   
0       Aves  ...  Critically Endangered           True    Not Sensitive   
..       ...  ...                    ...            ...              ...   
0      Flora  ...             Vulnerable          False    Not Sensitive   
0      Flora  ...  Critically Endangered          False    Not Sensitive   
0      Flora  ...  Critically Endangered          False    Not Sensitive   
0      Flora  ...             Endangered          False    Not Sensitive   
0      Flora  ...  Critically Endangered          False    No

In [139]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['bionetTaxonID','name','status','bionet_geoprivacy','kingdom','class','family', 'lsid']],
                    conservationlist[['bionetTaxonID','name','status','bionet_geoprivacy','kingdom','class','family', 'lsid']]]).drop_duplicates()
statelist

bionetTaxonID                           name                 status  \
0        10975.0       Callocephalon fimbriatum             Vulnerable   
0        10562.0                  Ninox strenua             Vulnerable   
0        10821.0               Tyto tenebricosa             Vulnerable   
0        10561.0                Ninox connivens             Vulnerable   
0        10413.0      Hoplocephalus bungaroides             Endangered   
..           ...                            ...                    ...   
0        10547.0    Muehlenbeckia sp. Mt Norman             Vulnerable   
0        20375.0             Lobelia claviflora  Critically Endangered   
0        20376.0    Fontainea sp. Coffs Harbour  Critically Endangered   
0        10521.0           Melichrus gibberagee             Endangered   
0        20378.0  Pittosporum sp. Coffs Harbour  Critically Endangered   

   bionet_geoprivacy   kingdom     class          family  \
0           obscured  Animalia      Aves      Cacatuidae   
0           obscured  Animalia      Aves       Strigidae   
0           obscured  Animalia      Aves       Tytonidae   
0           obscured  Animalia      Aves       Strigidae   
0           obscured  Animalia  Reptilia        Elapidae   
..               ...       ...       ...             ...   
0               open   Plantae     Flora    Polygonaceae   
0               open   Plantae     Flora   Campanulaceae   
0               open   Plantae     Flora   Euphorbiaceae   
0               open   Plantae     Flora       Ericaceae   
0               open   Plantae     Flora  Pittosporaceae   

                                                 lsid  
0   https://biodiversity.org.au/afd/taxa/6c646af8-...  
0   https://biodiversity.org.au/afd/taxa/d1c5dee0-...  
0   https://biodiversity.org.au/afd/taxa/645b287c-...  
0   https://biodiversity.org.au/afd/taxa/bd332ca4-...  
0   https://biodiversity.org.au/afd/taxa/f579a6fc-...  
..                                                ...  
0                                      ALA_DR650_1014  
0                                      ALA_DR650_1015  
0                                      ALA_DR650_1016  
0                                      ALA_DR650_1017  
0                                      ALA_DR650_1018  

[1024 rows x 8 columns]

In [140]:
# check for duplicates with conflicting information
statelist.groupby('bionetTaxonID').filter(lambda x: len(x) > 1)#.sort('size',ascending=False)
#df.groupby('hash').filter(lambda group: len(group) > 1).sort('size', ascending=False)

bionetTaxonID              name                 status bionet_geoprivacy  \
0       10840.0  Wollemia nobilis  Critically Endangered          obscured   
0       10840.0  Wollemia nobilis  Critically Endangered           private   

   kingdom  class         family  \
0  Plantae  Flora  Araucariaceae   
0  Plantae  Flora  Araucariaceae   

                                               lsid  
0  https://id.biodiversity.org.au/node/apni/2899238  
0  https://id.biodiversity.org.au/node/apni/2899238

### 4. Equivalent IUCN statuses

In [141]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild', 'Extinct'}
statelist.groupby(['status'])['status'].count()

status
Critically Endangered    115
Endangered               426
Extinct                   72
Not Listed                 5
Vulnerable               406
Name: status, dtype: int64

In [142]:
iucnStatusMappings = {
    'Not Listed':'Not Evaluated',
}

### 5. Determine best place ID to use

In [143]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 6825

place_id  place_name       place_display_name 
6744      Australia        Australia                5
6825      New South Wales  New South Wales, AU    161
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [144]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['name','status','bionet_geoprivacy', 'lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='name',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['name'])

mergedstatuses

name                 status  \
649                Acacia acanthoclada             Endangered   
442                 Acacia acrionastes             Endangered   
141                       Acacia atrox             Endangered   
510                   Acacia ausfeldii             Vulnerable   
734                      Acacia bakeri             Vulnerable   
..                                 ...                    ...   
115                   Zieria parrisiae  Critically Endangered   
795                   Zieria prostrata             Endangered   
796                 Zieria tuberculata             Vulnerable   
816  Zosterops lateralis tephropleurus             Vulnerable   
814                 Zosterops strenuus                Extinct   

    bionet_geoprivacy                                               lsid  \
649              open   https://id.biodiversity.org.au/node/apni/2898796   
442              open   https://id.biodiversity.org.au/node/apni/2900521   
141          obscured   https://id.biodiversity.org.au/node/apni/2894816   
510              open   https://id.biodiversity.org.au/node/apni/2886531   
734              open   https://id.biodiversity.org.au/node/apni/2902312   
..                ...                                                ...   
115          obscured   https://id.biodiversity.org.au/node/apni/2919089   
795              open   https://id.biodiversity.org.au/node/apni/2915817   
796              open   https://id.biodiversity.org.au/node/apni/2901823   
816              open  https://biodiversity.org.au/afd/taxa/fdbd023e-...   
814              open  https://biodiversity.org.au/afd/taxa/942564ee-...   

    status_id    scientificName taxon_id  user_id description iucn  \
649       NaN               NaN      NaN      NaN         NaN  NaN   
442       NaN               NaN      NaN      NaN         NaN  NaN   
141    152295      Acacia atrox   898643   708886               40   
510       NaN               NaN      NaN      NaN         NaN  NaN   
734       NaN               NaN      NaN      NaN         NaN  NaN   
..        ...               ...      ...      ...         ...  ...   
115       NaN               NaN      NaN      NaN         NaN  NaN   
795    165911  Zieria prostrata   897289  3669610               40   
796       NaN               NaN      NaN      NaN         NaN  NaN   
816       NaN               NaN      NaN      NaN         NaN  NaN   
814       NaN               NaN      NaN      NaN         NaN  NaN   

                                authority status_inat geoprivacy place_id  \
649                                   NaN         NaN        NaN      NaN   
442                                   NaN         NaN        NaN      NaN   
141  NSW Office of Environment & Heritage  endangered   obscured     6825   
510                                   NaN         NaN        NaN      NaN   
734                                   NaN         NaN        NaN      NaN   
..                                    ...         ...        ...      ...   
115                                   NaN         NaN        NaN      NaN   
795  NSW Office of Environment & Heritage          EN                6825   
796                                   NaN         NaN        NaN      NaN   
816                                   NaN         NaN        NaN      NaN   
814                                   NaN         NaN        NaN      NaN   

      place_display_name  
649                  NaN  
442                  NaN  
141  New South Wales, AU  
510                  NaN  
734                  NaN  
..                   ...  
115                  NaN  
795  New South Wales, AU  
796                  NaN  
816                  NaN  
814                  NaN  

[1027 rows x 15 columns]

In [145]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "New South Wales Office of Environment and Heritage"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://www.environment.nsw.gov.au/threatenedspeciesapp"
mergedstatuses['new_url'] = mergedstatuses['lsid']
mergedstatuses.rename(columns={'bionet_geoprivacy':'new_geoprivacy'},inplace=True)
mergedstatuses['new_place_id'] = '6825'  # NEW SOUTH WALES
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Threatened')
mergedstatuses

name                 status new_geoprivacy  \
649                Acacia acanthoclada             Endangered           open   
442                 Acacia acrionastes             Endangered           open   
141                       Acacia atrox             Endangered       obscured   
510                   Acacia ausfeldii             Vulnerable           open   
734                      Acacia bakeri             Vulnerable           open   
..                                 ...                    ...            ...   
115                   Zieria parrisiae  Critically Endangered       obscured   
795                   Zieria prostrata             Endangered           open   
796                 Zieria tuberculata             Vulnerable           open   
816  Zosterops lateralis tephropleurus             Vulnerable           open   
814                 Zosterops strenuus                Extinct           open   

                                                  lsid status_id  \
649   https://id.biodiversity.org.au/node/apni/2898796       NaN   
442   https://id.biodiversity.org.au/node/apni/2900521       NaN   
141   https://id.biodiversity.org.au/node/apni/2894816    152295   
510   https://id.biodiversity.org.au/node/apni/2886531       NaN   
734   https://id.biodiversity.org.au/node/apni/2902312       NaN   
..                                                 ...       ...   
115   https://id.biodiversity.org.au/node/apni/2919089       NaN   
795   https://id.biodiversity.org.au/node/apni/2915817    165911   
796   https://id.biodiversity.org.au/node/apni/2901823       NaN   
816  https://biodiversity.org.au/afd/taxa/fdbd023e-...       NaN   
814  https://biodiversity.org.au/afd/taxa/942564ee-...       NaN   

       scientificName taxon_id  user_id description iucn  ... geoprivacy  \
649               NaN      NaN      NaN         NaN  NaN  ...        NaN   
442               NaN      NaN      NaN         NaN  NaN  ...        NaN   
141      Acacia atrox   898643   708886               40  ...   obscured   
510               NaN      NaN      NaN         NaN  NaN  ...        NaN   
734               NaN      NaN      NaN         NaN  NaN  ...        NaN   
..                ...      ...      ...         ...  ...  ...        ...   
115               NaN      NaN      NaN         NaN  NaN  ...        NaN   
795  Zieria prostrata   897289  3669610               40  ...              
796               NaN      NaN      NaN         NaN  NaN  ...        NaN   
816               NaN      NaN      NaN         NaN  NaN  ...        NaN   
814               NaN      NaN      NaN         NaN  NaN  ...        NaN   

    place_id   place_display_name  \
649      NaN                  NaN   
442      NaN                  NaN   
141     6825  New South Wales, AU   
510      NaN                  NaN   
734      NaN                  NaN   
..       ...                  ...   
115      NaN                  NaN   
795     6825  New South Wales, AU   
796      NaN                  NaN   
816      NaN                  NaN   
814      NaN                  NaN   

                                         new_authority  \
649  New South Wales Office of Environment and Heri...   
442  New South Wales Office of Environment and Heri...   
141  New South Wales Office of Environment and Heri...   
510  New South Wales Office of Environment and Heri...   
734  New South Wales Office of Environment and Heri...   
..                                                 ...   
115  New South Wales Office of Environment and Heri...   
795  New South Wales Office of Environment and Heri...   
796  New South Wales Office of Environment and Heri...   
816  New South Wales Office of Environment and Heri...   
814  New South Wales Office of Environment and Heri...   

                                       new_description  \
649  Listed as Threatened - refer to https://www.en...   
442  Listed as Threatened - refer to https://www.en...   
141  Listed as Threatened -

## Updates

In [146]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('name')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','name','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'name':'taxon_name',
                                  'status_id':'id'})
updates

action             taxon_name      id taxon_id                 status  \
141  UPDATE           Acacia atrox  152295   898643             Endangered   
624  UPDATE   Acacia baueri aspera  159920  1111698             Vulnerable   
97   UPDATE     Acacia dangarensis  152286   775137  Critically Endangered   
609  UPDATE       Acacia pubifolia  162893   897779             Endangered   
448  UPDATE          Acacia ruppii  162887   578518             Endangered   
..      ...                    ...     ...      ...                    ...   
112  UPDATE  Viola cleistogamoides  152321   566603             Endangered   
877  UPDATE       Wollemia nobilis  152312    49381  Critically Endangered   
144  UPDATE       Wollemia nobilis  152312    49381  Critically Endangered   
705  UPDATE     Zieria involucrata  266172   604220             Endangered   
795  UPDATE       Zieria prostrata  165911   897289             Endangered   

    iucn_equivalent                                          authority  \
141      Vulnerable  New South Wales Office of Environment and Heri...   
624      Vulnerable  New South Wales Office of Environment and Heri...   
97       Vulnerable  New South Wales Office of Environment and Heri...   
609      Vulnerable  New South Wales Office of Environment and Heri...   
448      Vulnerable  New South Wales Office of Environment and Heri...   
..              ...                                                ...   
112      Vulnerable  New South Wales Office of Environment and Heri...   
877      Vulnerable  New South Wales Office of Environment and Heri...   
144      Vulnerable  New South Wales Office of Environment and Heri...   
705      Vulnerable  New South Wales Office of Environment and Heri...   
795      Vulnerable  New South Wales Office of Environment and Heri...   

                                                   url geoprivacy place_id  \
141   https://id.biodiversity.org.au/node/apni/2894816   obscured     6825   
624   https://id.biodiversity.org.au/node/apni/2892219       open     6825   
97    https://id.biodiversity.org.au/node/apni/2915433   obscured     6825   
609  https://id.biodiversity.org.au/taxon/apni/5129...       open     6825   
448   https://id.biodiversity.org.au/node/apni/2901189       open     6825   
..                                                 ...        ...      ...   
112   https://id.biodiversity.org.au/node/apni/2892487   obscured     6825   
877   https://id.biodiversity.org.au/node/apni/2899238    private     6825   
144   https://id.biodiversity.org.au/node/apni/2899238   obscured     6825   
705   https://id.biodiversity.org.au/node/apni/2888006       open     6825   
795   https://id.biodiversity.org.au/node/apni/2915817       open     6825   

      username                                        description  
141  peggydnew  Listed as Threatened - refer to https://www.en...  
624  peggydnew  Listed as Threatened - refer to https://www.en...  
97   peggydnew  Listed as Threatened - refer to https://www.en...  
609  peggydnew  Listed as Threatened - refer to https://www.en...  
448  peggydnew  Listed as Threatened - refer to https://www.en...  
..         ...                                                ...  
112  peggydnew  Listed as Threatened - refer to https://www.en...  
877  peggydnew  Listed as Threatened - refer to https://www.en...  
144  peggydnew  Listed as Threatened - refer to https://www.en...  
705  peggydnew  Listed as Threatened - refer to https://www.en...  
795  peggydnew  Listed as Threatened - refer to https://www.en...  

[150 rows x 12 columns]

## No status in iNaturalist via straight scientificName match
The NSW records that didn't match up to a status in iNaturalist

In [147]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="name",right_on="scientificName")
noinatstatus

name                 status new_geoprivacy  \
0                  Acacia acanthoclada             Endangered           open   
1                   Acacia acrionastes             Endangered           open   
2                     Acacia ausfeldii             Vulnerable           open   
3                        Acacia bakeri             Vulnerable           open   
4                      Acacia bynoeana             Endangered           open   
..                                 ...                    ...            ...   
872     Zieria odorifera warrabahensis  Critically Endangered           open   
873                   Zieria parrisiae  Critically Endangered       obscured   
874                 Zieria tuberculata             Vulnerable           open   
875  Zosterops lateralis tephropleurus             Vulnerable           open   
876                 Zosterops strenuus                Extinct           open   

                                                  lsid status_id  \
0     https://id.biodiversity.org.au/node/apni/2898796       NaN   
1     https://id.biodiversity.org.au/node/apni/2900521       NaN   
2     https://id.biodiversity.org.au/node/apni/2886531       NaN   
3     https://id.biodiversity.org.au/node/apni/2902312       NaN   
4     https://id.biodiversity.org.au/node/apni/2897229       NaN   
..                                                 ...       ...   
872   https://id.biodiversity.org.au/node/apni/2914041       NaN   
873   https://id.biodiversity.org.au/node/apni/2919089       NaN   
874   https://id.biodiversity.org.au/node/apni/2901823       NaN   
875  https://biodiversity.org.au/afd/taxa/fdbd023e-...       NaN   
876  https://biodiversity.org.au/afd/taxa/942564ee-...       NaN   

    scientificName_x taxon_id user_id description iucn  ...          class  \
0                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
1                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
2                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
3                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
4                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
..               ...      ...     ...         ...  ...  ...            ...   
872              NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
873              NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
874              NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
875              NaN      NaN     NaN         NaN  NaN  ...           Aves   
876              NaN      NaN     NaN         NaN  NaN  ...           Aves   

             order        family      genus specificEpithet  \
0          Fabales      Fabaceae     Acacia    acanthoclada   
1          Fabales      Fabaceae     Acacia     acrionastes   
2          Fabales      Fabaceae     Acacia       ausfeldii   
3          Fabales      Fabaceae     Acacia          bakeri   
4          Fabales      Fabaceae     Acacia        bynoeana   
..             ...           ...        ...             ...   
872     Sapindales      Rutaceae     Zieria       odorifera   
873     Sapindales      Rutaceae     Zieria       parrisiae   
874     Sapindales      Rutaceae     Zieria     tuberculata   
875  Passeriformes  Zosteropidae  Zosterops       lateralis   
876  Passeriformes  Zosteropidae  Zosterops        strenuus   

    infraspecificEpithet              modified  \
0                    NaN  2022-08-22T22:51:08Z   
1                    NaN  2022-04-06T22:12:50Z   
2                    NaN  2022-06-07T09:01:00Z   
3                    NaN  2022-04-07T01:35:47Z   
4                    NaN  2022-04-07T01:36:15Z   
..                   ...                   ...   
872        warrabahensis  2021-07-28T04:47:59Z   
873                  NaN  2020-02-19T19:35:12Z   
874                  NaN  2021-07-28T04:47:52Z   
875        tephropleurus  2020-01-16T00:05:01Z   
876         

In [148]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

name                 status new_geoprivacy  \
0                  Acacia acanthoclada             Endangered           open   
1                   Acacia acrionastes             Endangered           open   
2                     Acacia ausfeldii             Vulnerable           open   
3                        Acacia bakeri             Vulnerable           open   
4                      Acacia bynoeana             Endangered           open   
..                                 ...                    ...            ...   
872     Zieria odorifera warrabahensis  Critically Endangered           open   
873                   Zieria parrisiae  Critically Endangered       obscured   
874                 Zieria tuberculata             Vulnerable           open   
875  Zosterops lateralis tephropleurus             Vulnerable           open   
876                 Zosterops strenuus                Extinct           open   

                                                  lsid status_id  \
0     https://id.biodiversity.org.au/node/apni/2898796       NaN   
1     https://id.biodiversity.org.au/node/apni/2900521       NaN   
2     https://id.biodiversity.org.au/node/apni/2886531       NaN   
3     https://id.biodiversity.org.au/node/apni/2902312       NaN   
4     https://id.biodiversity.org.au/node/apni/2897229       NaN   
..                                                 ...       ...   
872   https://id.biodiversity.org.au/node/apni/2914041       NaN   
873   https://id.biodiversity.org.au/node/apni/2919089       NaN   
874   https://id.biodiversity.org.au/node/apni/2901823       NaN   
875  https://biodiversity.org.au/afd/taxa/fdbd023e-...       NaN   
876  https://biodiversity.org.au/afd/taxa/942564ee-...       NaN   

    scientificName_x taxon_id user_id description iucn  ...          class  \
0                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
1                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
2                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
3                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
4                NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
..               ...      ...     ...         ...  ...  ...            ...   
872              NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
873              NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
874              NaN      NaN     NaN         NaN  NaN  ...  Magnoliopsida   
875              NaN      NaN     NaN         NaN  NaN  ...           Aves   
876              NaN      NaN     NaN         NaN  NaN  ...           Aves   

             order        family      genus specificEpithet  \
0          Fabales      Fabaceae     Acacia    acanthoclada   
1          Fabales      Fabaceae     Acacia     acrionastes   
2          Fabales      Fabaceae     Acacia       ausfeldii   
3          Fabales      Fabaceae     Acacia          bakeri   
4          Fabales      Fabaceae     Acacia        bynoeana   
..             ...           ...        ...             ...   
872     Sapindales      Rutaceae     Zieria       odorifera   
873     Sapindales      Rutaceae     Zieria       parrisiae   
874     Sapindales      Rutaceae     Zieria     tuberculata   
875  Passeriformes  Zosteropidae  Zosterops       lateralis   
876  Passeriformes  Zosteropidae  Zosterops        strenuus   

    infraspecificEpithet              modified  \
0                    NaN  2022-08-22T22:51:08Z   
1                    NaN  2022-04-06T22:12:50Z   
2                    NaN  2022-06-07T09:01:00Z   
3                    NaN  2022-04-07T01:35:47Z   
4                    NaN  2022-04-07T01:36:15Z   
..                   ...                   ...   
872        warrabahensis  2021-07-28T04:47:59Z   
873                  NaN  2020-02-19T19:35:12Z   
874                  NaN  2021-07-28T04:47:52Z   
875        tephropleurus  2020-01-16T00:05:01Z   
876         

In [149]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['name'])
additions['action'] = 'ADD'
additions = additions[['action','name','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'name':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action                         taxon_name   id taxon_id  \
0      ADD                Acacia acanthoclada  NaN   787701   
1      ADD                 Acacia acrionastes  NaN   898576   
2      ADD                   Acacia ausfeldii  NaN   139906   
3      ADD                      Acacia bakeri  NaN   524410   
4      ADD                    Acacia bynoeana  NaN   543213   
..     ...                                ...  ...      ...   
872    ADD     Zieria odorifera warrabahensis  NaN  1247632   
873    ADD                   Zieria parrisiae  NaN   965722   
874    ADD                 Zieria tuberculata  NaN  1247618   
875    ADD  Zosterops lateralis tephropleurus  NaN   980046   
876    ADD                 Zosterops strenuus  NaN    17457   

                    status iucn_equivalent  \
0               Endangered      Vulnerable   
1               Endangered      Vulnerable   
2               Vulnerable      Vulnerable   
3               Vulnerable      Vulnerable   
4               Endangered      Vulnerable   
..                     ...             ...   
872  Critically Endangered      Vulnerable   
873  Critically Endangered      Vulnerable   
874             Vulnerable      Vulnerable   
875             Vulnerable      Vulnerable   
876                Extinct      Vulnerable   

                                             authority  \
0    New South Wales Office of Environment and Heri...   
1    New South Wales Office of Environment and Heri...   
2    New South Wales Office of Environment and Heri...   
3    New South Wales Office of Environment and Heri...   
4    New South Wales Office of Environment and Heri...   
..                                                 ...   
872  New South Wales Office of Environment and Heri...   
873  New South Wales Office of Environment and Heri...   
874  New South Wales Office of Environment and Heri...   
875  New South Wales Office of Environment and Heri...   
876  New South Wales Office of Environment and Heri...   

                                                   url geoprivacy place_id  \
0     https://id.biodiversity.org.au/node/apni/2898796       open     6825   
1     https://id.biodiversity.org.au/node/apni/2900521       open     6825   
2     https://id.biodiversity.org.au/node/apni/2886531       open     6825   
3     https://id.biodiversity.org.au/node/apni/2902312       open     6825   
4     https://id.biodiversity.org.au/node/apni/2897229       open     6825   
..                                                 ...        ...      ...   
872   https://id.biodiversity.org.au/node/apni/2914041       open     6825   
873   https://id.biodiversity.org.au/node/apni/2919089   obscured     6825   
874   https://id.biodiversity.org.au/node/apni/2901823       open     6825   
875  https://biodiversity.org.au/afd/taxa/fdbd023e-...       open     6825   
876  https://biodiversity.org.au/afd/taxa/942564ee-...       open     6825   

      username                                        description  
0    peggydnew  Listed as Threatened - refer to https://www.en...  
1    peggydnew  Listed as Threatened - refer to https://www.en...  
2    peggydnew  Listed as Threatened - refer to https://www.en...  
3    peggydnew  Listed as Threatened - refer to https://www.en...  
4    peggydnew  Listed as Threatened - refer to https://www.en...  
..         ...                                                ...  
872  peggydnew  Listed as Threatened - refer to https://www.en...  
873  peggydnew  Listed as Threatened - refer to https://www.en...  
874  peggydnew  Listed as Threatened - refer to https://www.en...  
875  peggydnew  Listed as Threatened - refer to https://www.en...  
876  peggydnew  Listed as Threatened - refer to https://www.en...  

[731 rows x 12 columns]

In [150]:
# write these to the file
pd.concat([updates,additions]).to_csv(sourcedir + "nsw.csv", index=False)

In [151]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

name      status new_geoprivacy  \
23      Acacia terminalis Eastern Sydney  Endangered           open   
37   Amperea xiphoclada var. pedicellata     Extinct           open   
45                        Angophora exul  Endangered       obscured   
50                    Antaresia stimsoni  Vulnerable           open   
76          Astrotricha sp. Wallagaraugh  Endangered           open   
..                                   ...         ...            ...   
827              Trianthema cypseleoides     Extinct           open   
833                   Tylophora linearis  Vulnerable           open   
834                   Tylophora woollsii  Endangered           open   
838           Typhonium sp. aff. brownii  Endangered       obscured   
843                   Velleia perfoliata  Vulnerable           open   

                                                  lsid status_id  \
23                                      ALA_DR650_1011       NaN   
37    https://id.biodiversity.org.au/node/apni/2910887       NaN   
45    https://id.biodiversity.org.au/node/apni/2913071       NaN   
50   https://biodiversity.org.au/afd/taxa/e87421fb-...       NaN   
76                                       ALA_DR650_908       NaN   
..                                                 ...       ...   
827   https://id.biodiversity.org.au/node/apni/2916288       NaN   
833   https://id.biodiversity.org.au/node/apni/2906372       NaN   
834   https://id.biodiversity.org.au/node/apni/2901871       NaN   
838                                      ALA_DR487_133       NaN   
843   https://id.biodiversity.org.au/node/apni/2920937       NaN   

    scientificName_x taxon_id user_id description iucn  ... class order  \
23               NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
37               NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
45               NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
50               NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
76               NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
..               ...      ...     ...         ...  ...  ...   ...   ...   
827              NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
833              NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
834              NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
838              NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   
843              NaN      NaN     NaN         NaN  NaN  ...   NaN   NaN   

    family genus specificEpithet infraspecificEpithet modified  \
23     NaN   NaN             NaN                  NaN      NaN   
37     NaN   NaN             NaN                  NaN      NaN   
45     NaN   NaN             NaN                  NaN      NaN   
50     NaN   NaN             NaN                  NaN      NaN   
76     NaN   NaN             NaN                  NaN      NaN   
..     ...   ...             ...                  ...      ...   
827    NaN   NaN             NaN                  NaN      NaN   
833    NaN   NaN             NaN                  NaN      NaN   
834    NaN   NaN             NaN                  NaN      NaN   
838    NaN   NaN             NaN                  NaN      NaN   
843    NaN   NaN             NaN                  NaN      NaN   

    scientificName_y taxonRank references  
23               NaN       NaN        NaN  
37               NaN       NaN        NaN  
45               NaN       NaN        NaN  
50               NaN       NaN        NaN  
76               NaN       NaN        NaN  
..               ...       ...        ...  
827              NaN       NaN        NaN  
833              NaN       NaN        NaN  
834              NaN       NaN        NaN  
838              NaN       NaN        NaN  
843              NaN       NaN        NaN  

[146 rows x 38 columns]

### are there any that need to be removed?
nsw sensitive list count: 198
nsw inat statuses count: 166

updates to inat status: 150
additional inat status: 732
nsw statuses we can't find a taxon match for in iNaturalist: 149
total: 1395 (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over: 166-150=16 that may need checking

In [152]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]


status_id taxon_id  user_id place_id source_id  \
477     264941  1061113  3669610     6825             
681     164757  1098159    58320     6825             
1147    152307   116844   708886     6825     16650   
736     166035  1194478  3669610     6825             
737     166036  1194478  3669610     6744             
51      167249  1227491  3669610     6825             
397     264070  1255425  3669610     6825             
809     262163  1289564   708886     6825     16650   
3367    152252    19111   708886     6825     16650   
2653    164637    19275    84719     6825             
2608    164300   369148   516268     6825             
1130    152282   505120   708886     6825     16650   
65      168033   634763   702203     6825             
1137    152293   775274   708886     6825     16650   
3426    152239   792286   708886     6825     16650   
2437    165525   906591    58320     6825             
836     155144   966856   708886     6825     16650   

                                              authority  \
477   New South Wales Biodiversity Conservation Act ...   
681                NSW Office of Environment & Heritage   
1147               NSW Office of Environment & Heritage   
736                NSW Office of Environment & Heritage   
737                               Australian Government   
51                 NSW Office of Environment & Heritage   
397                NSW Office of Environment & Heritage   
809                NSW Office of Environment & Heritage   
3367               NSW Office of Environment & Heritage   
2653                                      IUCN Red List   
2608        NSW Threatened Species Scientific Committee   
1130               NSW Office of Environment & Heritage   
65               Threatened Species Protection Act 1995   
1137               NSW Office of Environment & Heritage   
3426               NSW Office of Environment & Heritage   
2437               NSW Office of Environment & Heritage   
836                NSW Office of Environment & Heritage   

                     status  \
477                      EX   
681              Vulnerable   
1147  critically endangered   
736                      EN   
737                      EN   
51                       EN   
397                      CR   
809              vulnerable   
3367             vulnerable   
2653                     EN   
2608                     EN   
1130             vulnerable   
65                       EN   
1137  endangered population   
3426             vulnerable   
2437             Endangered   
836              endangered   

                                                    url  \
477   https://bie.ala.org.au/species/https://id.biod...   
681       https://www.inaturalist.org/taxa/1098159/edit   
1147  https://www.environment.nsw.gov.au/resources/a...   
736   https://www.environment.nsw.gov.au/threateneds...   
737   https://www.environment.nsw.gov.au/threateneds...   
51    https://www.environment.nsw.gov.au/threateneds...   
397   https://www.environment.nsw.gov.au/threateneds...   
809   https://www.environment.nsw.gov.au/resources/a...   
3367  https://www.environment.nsw.gov.au/resources/a...   
2653  https://www.iucnredlist.org/species/22727593/1...   
2608  https://www.dpi.nsw.gov.au/__data/assets/pdf_f...   
1130  https://www.environment.nsw.gov.au/resources/a...   
65    http://www.environment.gov.au/resource/nationa...   
1137  https://www.environment.nsw.gov.au/resources/a...   
3426  https://www.environment.nsw.gov.au/resources/a...   
2437  https://www.environment.nsw.gov.au/threatenedS...   
836   https://www.environment.nsw.gov.au/resources/a...   

                                            description geoprivacy  ...  \
477                                    Presumed Extinct       open  ...   
681                                                                 ...   
1147                                                      obscured  ...   
736                          